File: LeetCode/80RemoveDuplicatesFromSortedArrayII/RemoveDuplicatesFromSortedArrayII.py
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        # Partition method -> Time: O(n), Space: O(1)
        k = 0
        d = {}
        for i in range(len(nums)):
            if nums[i] not in d:
                d[nums[i]] = 1
                nums[k] = nums[i]
                k += 1
            elif d[nums[i]] < 2:
                d[nums[i]] += 1
                nums[k] = nums[i]
                k += 1
        return k

File: LeetCode/80RemoveDuplicatesFromSortedArrayII/RemoveDuplicatesFromSortedArrayII.cpp
class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        // Partition method -> Time: O(n), Space: O(1)
        int k = 0;
        unordered_map<int, int> m;
        for (int i = 0; i < nums.size(); i++) {
            if (m.count(nums[i]) == 0) {
                nums[k] = nums[i];
                m[nums[i]] = 1;
                k++;
            }
            else if (m[nums[i]] == 1) {
                nums[k] = nums[i];
                m[nums[i]] = 2;
                k++;
            }
        }
        return k;
    }
};